### DISCUSSION TITLE
How to optimize Ocelot startup time?

### DISCUSSION THREAD

**Discussion by heizel184** [ThreeMammals/Ocelot#1857] [Q&A] [answered]
Hi I am using Ocelot 15.0.6 as a gateway. Since Ocelot supports multiple configuration files. Previously I include seven different configurations files with the routes of different projects API, The problem occurred when a new file is added with more than 80 routes in it. After adding that file The ocelot took too long at startup that the application got stuck and shut down later when deploying gateway.

The current behavior of ocelot is unacceptable, application should not take much time for a startup. How can I optimize that time or speed up the startup time of ocelot at production?

## Steps to Reproduce the Problem
Add multiple configuration files with a large no of routes
## Specifications

  - Version: 15.0.6

**Comment by ks1990cn** [comment]
@heizel184 are you still facing it on latest versions?

**Comment by raman-m** [comment]
@heizel184  Dear Hina!
Thanks for reporting it!

Your issue is quite outdated because of legacy version 15.0.6.
Could you upgrade your solution to the latest version and check the issue once again please?

----
> The problem occurred when a new file is added with more than 80 routes in it.

Are you sure that all files are valid? So, have all these files valid JSON formatting and syntax?
Can you attach all config to the issue plz?

-----
> After adding that file The ocelot took too long at startup that the application got stuck and shut down later when deploying gateway.

Deployment can fail because of critical error when starting up the app...
Can you attach Ocelot log please?

**Reply by raman-m** [reply, marked as answer]
@heizel184
> How to optimize Ocelot startup time?

Just keep all routes in one file!  
To be fair, I don't understand creation of a separate file for each route! It is strange. You could use another merging strategy: for example create separate file for each downstream host to keep all routes to this downstream service in one **.json** file. I believe your number of files will decrease in times. 
But we are working now on your user scenario when there are a lot of tiny files... We will try to check Ocelot behavior having 50-100 files in configuration. Please watch this PR: #1856

**Comment by raman-m** [comment]
@heizel184 
>## Steps to Reproduce the Problem
>Add multiple configuration files with a large no of routes

How many?
But it will be better if you attach those files!

**Comment by ks1990cn** [comment]
Following result based on #1964 

HeavyRoutesStartupBenchmark.StartOcelotWithLargeConfigurations: DefaultJob
Runtime = .NET 6.0.25 (6.0.2523.51912), X64 RyuJIT AVX2; GC = Concurrent Workstation
Mean = 7.135 s, StdErr = 0.061 s (0.85%), N = 92, StdDev = 0.581 s
Min = 5.827 s, Q1 = 6.815 s, Median = 6.959 s, Q3 = 7.477 s, Max = 8.815 s
IQR = 0.661 s, LowerFence = 5.823 s, UpperFence = 8.469 s
ConfidenceInterval = [6.929 s; 7.341 s] (CI 99.9%), Margin = 0.206 s (2.89% of Mean)
Skewness = 0.67, Kurtosis = 4.01, MValue = 2.3
-------------------- Histogram --------------------
[5.694 s ; 6.031 s) | @@@@
[6.031 s ; 6.373 s) |
[6.373 s ; 6.635 s) | @
[6.635 s ; 6.973 s) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[6.973 s ; 7.348 s) | @@@@@@@@@@@@@@@@@@
[7.348 s ; 7.626 s) | @@@@@@@
[7.626 s ; 7.964 s) | @@@@@@@@@@@@
[7.964 s ; 8.275 s) | @@@
[8.275 s ; 8.617 s) | @
[8.617 s ; 8.984 s) | @@@
---------------------------------------------------

// * Summary *

BenchmarkDotNet v0.13.11, Windows 11 (10.0.22621.3155/22H2/2022Update/SunValley2)
Intel Core i7-8650U CPU 1.90GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.200-preview.23624.5
  [Host]     : .NET 6.0.25 (6.0.2523.51912), X64 RyuJIT AVX2 [AttachedDebugger]
  DefaultJob : .NET 6.0.25 (6.0.2523.51912), X64 RyuJIT AVX2


| Method                             | Mean    | Error    | StdDev   | StdErr   | Median  | Min     | Q1      | Q3      | Max     | Op/s   | Gen0       | Gen1      | Allocated |
|----------------------------------- |--------:|---------:|---------:|---------:|--------:|--------:|--------:|--------:|--------:|-------:|-----------:|----------:|----------:|
| StartOcelotWithLargeConfigurations | 7.135 s | 0.2059 s | 0.5808 s | 0.0605 s | 6.959 s | 5.827 s | 6.815 s | 7.477 s | 8.815 s | 0.1402 | 45000.0000 | 3000.0000 | 191.49 MB |

// * Warnings *
Environment
  Summary -> Benchmark was executed with attached debugger

// * Hints *
Outliers
  HeavyRoutesStartupBenchmark.StartOcelotWithLargeConfigurations: Default -> 8 outliers were removed (8.98 s..14.23 s)

// * Legends *
  Mean      : Arithmetic mean of all measurements
  Error     : Half of 99.9% confidence interval
  StdDev    : Standard deviation of all measurements
  StdErr    : Standard error of all measurements
  Median    : Value separating the higher half of all measurements (50th percentile)
  Min       : Minimum
  Q1        : Quartile 1 (25th percentile)
  Q3        : Quartile 3 (75th percentile)
  Max       : Maximum
  Op/s      : Operation per second
  Gen0      : GC Generation 0 collects per 1000 operations
  Gen1      : GC Generation 1 collects per 1000 operations
  Allocated : Allocated memory per single operation (managed only, inclusive, 1KB = 1024B)
  1 s       : 1 Second (1 sec)

// * Diagnostic Output - MemoryDiagnoser *


// ***** BenchmarkRunner: End *****
Run time: 00:15:35 (935.39 sec), executed benchmarks: 1

Global total time: 00:15:42 (942.7 sec), executed benchmarks: 1
// * Artifacts cleanup *
Artifacts cleanup is finished
